Added content about handling alerts and best practices regarding Logging-Sentry (#182)
axelerant/engg-handbook: content/docs/how-we-work/baseline/logging-details.md

---
title: Logging

> weight: 2

Reviewer: ```suggestion
weight: 13
```

I know there are no 13 pages under this section but in case we add more sub-pages, I would like it to maintain the order we have for the headings on the baseline page. Setting the weight to match that order right now makes it easier for us later.

description: We can't fix what we can't see. Logging is the most fundamental aspect of observability in how our applications run so that we fix a problem, even proactively. For this, we rely on something more sophisticated than simply logging into the database or syslog. Our tool of choice for this purpose is Sentry.
---

# Logging

We can't fix what we can't see. Logging is the most fundamental aspect of observability in how our applications run so that we fix a problem, even proactively. For this, we rely on something more sophisticated than simply logging into the database or syslog. Our tool of choice for this purpose is [Sentry](https://axelerant.sentry.io/).

> Any application that you build must be configured to write logs to Sentry. This is fairly straightforward as Sentry has SDKs in a number of languages and frameworks. For Drupal, use the [Raven](https://www.drupal.org/project/raven) module to pull in the integration and also configure what should be logged.

Reviewer: Since we have moved everything here, let's remove all the text following this from the main baseline page. That page should only have these paragraphs and make the word "Logging" in the first paragraph link to this page.

3. **Create a Ticket:**
   - **Manually Creating a Ticket:**
     - Create an Issue** in the alert details view to manually create a ticket by writing all relevant information provided by Sentry to ensure the issue is well-documented.

   - **Automatically Creating Tickets:**
     - Configure automatic ticket creation in **Project Settings** > **Alerts** by setting up alert rules to trigger ticket creation by specifying the conditions (e.g., error type, frequency) and actions (create a Jira ticket) for automatic ticket creation.

4. **Managing Tickets in Jira:**
   - Ensure tickets are assigned to the appropriate team or individual in Jira. Use Jira’s workflow management tools to track the progress of the ticket from creation to resolution.
   - **Tracking and Reporting:** Use Jira’s reporting features to track the status of logged issues. Generate reports to identify trends, recurring issues, and areas for improvement.
   - **Comments and Updates:** Utilize Jira’s commenting system to add updates, discuss potential solutions, and collaborate with team members directly within the ticket.

> ## Best Practices

Reviewer: Let's trim down "Best Practices" and rewrite it as a checklist. All of the below can be reduced to a checklist of 10 or so items.

1. **User Interactions:**
   - Use Drupal's built-in logging functions such as `watchdog` to log significant user actions and interactions.
   - Implement custom logging in custom modules using the `\Drupal::logger('custom_module')->info($message);` syntax to capture specific user events.

2. **Database Queries:**
   - Enable query logging in Drupal by configuring the database logging settings in `settings.php`.
   - Use the Database Logging module (`dblog`) to log and review database queries within the Drupal admin interface.
   - Implement custom logging for complex or critical database operations using hooks and custom logging functions.

3. **API Calls:**
   - Log all outgoing and incoming API requests and responses. Use the Guzzle middleware to log API requests made through Guzzle HTTP client.
   - Implement logging in custom API integrations by capturing request and response data and logging it using `\Drupal::logger`.

4. **Third-Party Integrations:**
   - Ensure that all third-party services and integrations have appropriate logging in place. This can include logging successful and failed requests, responses, and any relevant data exchanges.

>    - Utilize existing logging features provided by third-party modules and extend them as needed to capture additional details.

Reviewer: What do you mean by third-party integrations here? From what I understand, this is the same as the API calls above. If you are talking about using API to talk to third-party services, then maybe we just club these two.

5. **Error Levels:**

>    - Use appropriate error levels to categorize logs, but avoid sending debug and info logs to Sentry as they are not suitable for these log types.

Reviewer: ```suggestion
   - Sentry is an issue detection and reporting tool and does not need `debug` and `info` log messages. While you should log all messages to logging systems like `syslog` or external ones such as New Relic, only errors are relevant to Sentry.
```

>    - Set the logging threshold to `Warning`, `Error`, or `Critical` to filter and prioritize issues effectively.

Reviewer: ```suggestion
   - Set the logging threshold to `Error` or `Critical` to filter and prioritize issues effectively.
```

Let's not explicitly suggest "Warning". They may do that if they see a need.

6. **Monitoring and Review:**
   - Regularly review and monitor logs using the Drupal logging interface or integrate with external monitoring tools like Sentry to ensure comprehensive coverage.
   - Set up automated log review and alerting processes to quickly identify and address any logging gaps or issues.

### Regularly Review Logs

1. **Log Review Sessions:**
   - Schedule regular log review sessions to periodically review logs and identify recurring issues as part of your development and maintenance processes.

2. **Log Retention Policies:**

   - Implement log retention policies to manage log storage efficiently. Ensure that logs are archived and accessible for future reference.
   - Define log retention periods and configure log rotation to prevent excessive log storage and maintain performance.

### Optimize Alert Rules

1. **Alert Configuration:**
   - Fine-tune alert rules to minimize noise and focus on critical issues by setting appropriate thresholds and conditions for alerts.

2. **Alert Tuning:**

>    - Continuously monitor and adjust alert rules based on feedback and changing application behavior. Engage with the team to refine alert settings.

Reviewer: Isn't this already covered earlier?

Author: @hussainweb You are right, but it is mentioned here as a best practice.

### Automate Ticket Creation

1. **Integration:**
   - Automate the ticket creation process to streamline incident management. Use Sentry’s integration capabilities to automatically create tickets for high-priority issues.

2. **Automation Tools:**

>    - Explore automation tools and scripts that can help manage alerts and tickets more efficiently. This can include auto-assigning tickets, updating statuses, and notifying relevant stakeholders.

Reviewer: This was also covered before, right?

Author: @hussainweb You are right, but it is mentioned here as a best practice.

3. **Create a Ticket:**
   - **Manually Creating a Ticket:**
     - **Create an Issue** in the alert details view to manually create a ticket by writing all relevant information provided by Sentry to ensure the issue is well-documented.

   - **Automatically Creating Tickets:**
     - Configure automatic ticket creation in **Project Settings** > **Alerts** by setting up alert rules to trigger ticket creation by specifying the conditions (e.g., error type, frequency) and actions (create a Jira ticket) for automatic ticket creation.

4. **Managing Tickets in Jira:**
   - Ensure tickets are assigned to the appropriate team or individual in Jira. Use Jira’s workflow management tools to track the progress of the ticket from creation to resolution.
   - **Tracking and Reporting:** Use Jira’s reporting features to track the status of logged issues. Generate reports to identify trends, recurring issues, and areas for improvement.
   - **Comments and Updates:** Utilize Jira’s commenting system to add updates, discuss potential solutions, and collaborate with team members directly within the ticket.

> ## Best Practices
> 
> We apply these best practices to ensure effective logging, monitoring, and incident management in our Drupal applications:

Reviewer: ```suggestion
## Things to Remember

We apply these practices to ensure effective logging, monitoring, and incident management in our Drupal applications:
```

- Log significant user actions and interactions using Drupal's built-in functions and custom modules.
- Enable and review database query logging using `settings.php` and the Database Logging module (`dblog`).

> - Log all API requests, responses, and third-party service interactions using `Guzzle` middleware and capturing successful and failed requests.

Reviewer: Let's not mention logging *all* requests. It's not worth logging everything and we should not encourage that.

> - Set the logging threshold to Error or Critical to filter and prioritize issues effectively and avoid logging debug and info logs to Sentry.

Reviewer: ```suggestion
- For Sentry, set the logging threshold to Error or Critical to filter and prioritize issues effectively and avoid sending debug and info messages.
```

> - Schedule regular log review sessions to identify and address recurring issues.

Reviewer: ```suggestion
- Schedule regular log review sessions with the team to identify and address recurring issues.
```

> - Implement log retention policies and configure log rotation for efficient log storage and performance.

Reviewer: I am not too sure about this either. This is something that the infrastructure teams worry about. Developers don't really worry about log retention.

- Fine-tune alert rules to minimize noise and focus on critical issues.
- Continuously monitor and adjust alert rules based on feedback and changing application behavior.
- Use Sentry’s integration capabilities to automatically create tickets for high-priority issues.
- Explore automation tools to manage alerts and tickets, including auto-assigning and status updates.

> - Foster collaboration between development, operations, and support teams using Sentry’s collaboration features.

Reviewer: I am not too sure about this. I would rather that the team use a single place where they are comfortable to collaborate.
